### ISSUE TITLE
Next.js removes css modules causing unstyled pages in production

### ISSUE THREAD

**Issue by jantimon** [vercel/next.js#68328]
### Link to the code that reproduces this issue

https://codesandbox.io/p/devbox/gracious-keller-qdj5ld?file=%2Fpages%2Findex.tsx%3A1%2C29

### To Reproduce

1. Open a page that shows a component which is **normally loaded and lazy loaded** at the same time.
2. Use the pages router to do a client-side navigation to a page where the component is **only lazy loaded**.
3. The CSS is missing and the lazy loaded component is shown unstyled

Reproduction steps using the provided CodeSandbox:

1. Open the reproduction: https://codesandbox.io/p/devbox/gracious-keller-qdj5ld
2. Observe that the lazy-loaded Star component on the home page is **styled** 
![styled star](https://github.com/user-attachments/assets/904a9b9e-4b03-4961-b79a-ba8c761a8931)
3. Navigate to the about page (/about-us)
4. Reload
5. Click on the "Back to home" link to navigate to /
6. Observe that the lazy-loaded Star component on the home page is **unstyled**
![unstyled star](https://github.com/user-attachments/assets/037d60d6-e407-41b5-bdbc-099fbd8a1b81)

Video:

https://github.com/user-attachments/assets/2d528874-d1fd-4389-b4c1-1332ef74fa3d

### Current vs. Expected behavior

Current behavior:

When navigating from a page with both normally loaded and lazy-loaded components to a page with only lazy-loaded components, the CSS for the lazy-loaded component is missing.

Expected behavior:

The CSS for lazy-loaded components should be properly applied regardless of the navigation path or the loading method on the previous page.

### Provide environment information

```bash
Operating System:
  Platform: linux
  Arch: x64
  Version: #1 SMP PREEMPT_DYNAMIC Sun Aug  6 20:05:33 UTC 2023
  Available memory (MB): 4102
  Available CPU cores: 2
Binaries:
  Node: 20.9.0
  npm: 9.8.1
  Yarn: 1.22.19
  pnpm: 8.10.2
Relevant Packages:
  next: 15.0.0-canary.91 // Latest available version is detected (15.0.0-canary.91).
  eslint-config-next: N/A
  react: 19.0.0-rc.0
  react-dom: 19.0.0-rc.0
  typescript: 5.3.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Lazy Loading, Pages Router

### Which stage(s) are affected? (Select all that apply)

next build (local), Vercel (Deployed)

### Additional context

- I could reproduce the issue in Next.js versions:
  - `next` 12.3.4
  - `next` 13.0.0
  - `next` 14.0.0
  - `next` 14.1.3
  - `next` 14.2.3
  - `next` 15.0.0-canary.91
- It happens only for the `pages` router - the `app` router seems to work fine
- Happens only for `npm run build` - `npm run dev` seems to work fine
- The issue seems to be related to the client-side rendering logic in `packages/next/src/client/index.tsx`
- This bug cause significant styling issues in production environments

**Comment by icyJoseph** [comment]
An reproduction repo on Stackblitz, https://stackblitz.com/edit/nextjs-6mb95w?file=pages%2Fabout.tsx,pages%2Findex.tsx,components%2Flazy.tsx, just cancel the command, and run, `npm i && npx next build && npx next start`

Removed a bunch of code not needed to reproduce the issue, and included `next/dynamic`

**Comment by icyJoseph** [comment]
So, the issue seems to, at least, be originating here, https://github.com/vercel/next.js/blob/cab4b653997b83ec3b80bb2f0138d59e5d9e8f5c/packages/next/src/client/index.tsx#L682-L697

Specifically here:

```js
styleTag.setAttribute('media', 'x')
```

<img width="488" alt="Screenshot 2024-07-31 at 09 05 55" src="https://github.com/user-attachments/assets/6fea788d-04d9-4ab7-a989-61f02aa6cfa3">

Changing to `media="all"`, brings back the styles, as expected... so the question is why does it believe that this href doesn't apply... 🤔

**Comment by samcx** [comment]
@jantimon Thanks for submitting an issue!

**Comment by icyJoseph** [comment]
> @jantimon Thank you for submitting an issue!
> 
> Is there a reason why you are not using `next/dynamic` for lazy loading components → https://nextjs.org/docs/app/building-your-application/optimizing/lazy-loading#nextdynamic?

The Stackblitz example I made does use next/dynamic though. Same problem.

**Comment by samcx** [comment]
@icyJoseph @jantimon It looks like this is related to this long-running issue → https://github.com/vercel/next.js/issues/33286.

Good news, we've actually started picking this up! → https://github.com/vercel/next.js/pull/68396

**Comment by jantimon** [comment]
@samcx - Good news

really really cool!

@samcx - why no `next/dynamic`:

`next/dynamic` works only for react components and adds some (tiny) overhead for client only components - but it actually uses the same `import()` syntax to load the dynamic piece under the hood so it shouldn't matter


@icyJoseph - `styleTag.setAttribute('media', 'x')`

There are two different systems running: the [extract text plugin](https://github.com/webpack-contrib/mini-css-extract-plugin) which adds code to lazy loads the css chunk and the nextjs router which adds/removes styles to keep the correct order and tries to apply only relevant styles for the current page - the problem seems to be that the mini-css-extract-plugin adds the lazy css only once but nextjs will remove it every time you go to another page

**Comment by max-degterev** [comment]
I am getting the same issue now with app router, some css modules simply not loading for certain routes. Happens on both dev and prod

**Comment by jantimon** [comment]
@max-degterev that is probably another bug - can you please create a dedicated issue?

**Comment by max-degterev** [comment]
I don't think i can isolate. It affects some classes in some modules, haven't been able to deduce any logic to it yet. If i simply copy-paste the rule block (have duplicate rules) it compiles correctly. Really weird. Started happening today.

**Comment by max-degterev** [comment]
![Screenshot 2024-09-28 at 8 06 09 PM](https://github.com/user-attachments/assets/d5ac0bd6-8461-4b82-9d87-2a70dd326ab2)
![Screenshot 2024-09-28 at 8 06 21 PM](https://github.com/user-attachments/assets/856e9735-f9f7-4b63-a4c7-078f0f7190d3)

Literally what the hell

**Comment by max-degterev** [comment]
The bug was introduced in 14.2.2, locking to 14.2.1 solves the issue

**Comment by max-degterev** [comment]
Yep literally adding `.nextjs {}` to the top of each css file "solves" the issue 🤣

**Comment by max-degterev** [comment]
Yep literally adding `.nextjs {}` to the top of each css file "solves" the issue 😄

**Comment by max-degterev** [comment]
@jantimon yep i can't create a reproducible case for this, I guess the bug will stay. 🤷 

To anyone who wants to use css modules - add an empty class name at the top!

**Comment by max-degterev** [comment]
@jantimon

Aha I've managed to trace the error.

Output v14.2.1:
![Screenshot 2024-09-28 at 10 00 02 PM](https://github.com/user-attachments/assets/f991ffa8-5837-49cb-ac3d-85cb3e300b8a)

Output v14.2.13:
![Screenshot 2024-09-28 at 9 58 55 PM](https://github.com/user-attachments/assets/e51fd07b-3c5f-438b-860b-3c2313cd6286)

There is a rogue class with a semicolon after the closing curly. Because nextjs doesn't guarantee class order any class that follows that rogue definition will not apply. This is a floating bug that will be very hard to trace. Ideally compiler should throw or strip these.

This little semicolon just cost me 4 hours.

I'm not gonna create a separate issue for this because frankly your issue creation rules are not very inviting to collaborate. I will patch this in my codebase, the rest is up to you guys.

**Comment by devjiwonchoi** [comment]
Hey everyone, sorry for the long delay, and thank you for your patience.

The fix has been landed at release [v15.0.4-canary.32](https://github.com/vercel/next.js/releases/tag/v15.0.4-canary.32), so feel free to try it, and let me know if it doesn't work for you. We are expecting to release this in the next release. Thank you!
